pmd-java/src/main/resources/category/java/performance.xml

```diff
@@ -72,49 +72,82 @@ sb.append('a');     // use this instead
           class="net.sourceforge.pmd.lang.rule.XPathRule"
           externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_performance.html#avoidarrayloops">
         <description>
-Instead of manually copying data between two arrays, use the efficient Arrays.copyOf or System.arraycopy method instead.
+Instead of manually copying data between two arrays, use the more efficient `Arrays.copyOf`
+or `System.arraycopy` method instead.
+
+To copy only part of the array, use `Arrays.copyOfRange` or `System.arraycopy`.
+
+If you want to copy/move elements inside the _same_ array (e.g. shift the elements), use `System.arraycopy`.
         </description>
         <priority>3</priority>
         <properties>
             <property name="version" value="2.0"/>
             <property name="xpath">
                 <value>
 <![CDATA[
-//Statement[(ForStatement or WhileStatement) and
-count(*//AssignmentOperator[@Image = '='])=1
-and
-*/Statement
-[
-./Block/BlockStatement/Statement/StatementExpression/PrimaryExpression
-/PrimaryPrefix/Name/../../PrimarySuffix/Expression
-[(PrimaryExpression or AdditiveExpression) and count
-(.//PrimaryPrefix/Name)=1]//PrimaryPrefix/Name/@Image
-and
-./Block/BlockStatement/Statement/StatementExpression/Expression/PrimaryExpression
-/PrimaryPrefix/Name/../../PrimarySuffix[count
-(..//PrimarySuffix)=1]/Expression[(PrimaryExpression
-or AdditiveExpression) and count(.//PrimaryPrefix/Name)=1]
-//PrimaryPrefix/Name/@Image
-]]
+//Statement
+  [(ForStatement or WhileStatement)]
+  [count(*/Statement//AssignmentOperator[@Image = '='])=1]
+  [*/Statement/Block/BlockStatement/Statement
+        /StatementExpression[AssignmentOperator[@Image = '=']]
+        (: LHS - left hand side :)
+        [PrimaryExpression[count(PrimarySuffix) = 1]/PrimarySuffix[@ArrayDereference = true()]
+        /Expression[PrimaryExpression or AdditiveExpression]
+                   [count(.//PrimaryPrefix/Name) = 1 or
+                    (count(.//PrimaryPrefix/Name) = 2 and .//PrimaryPrefix/Name/@Image = //VariableDeclaratorId[@Final = true()]/@Name)
```

Comment:
That's basically allowing variable access, if the variable is final. However, this search is not very accurate and doesn't use correct scoping rules. E.g. a variable with the same name in a previous method would be used instead of the local variable.
We could limit it to search for local vars in the same method (`parent::MethodDeclaration//VariableDeclarator`) but then we would miss field declarations. I think, the most common use case would be to declare a constant as a static final field and use it in the relative addressing of array index access.

Although this is not 100% accurate, I hope it is good enough for now.

Reply:
I personally like to declare every variable as final by default, not just "constants", so it doesn't really hold the same meaning. See #4183 for an example where this becomes important.